SpatialGDK/ExternalSchemaCodegen/Programs/Improbable.CodeGen.Unreal/Types.cs

```diff
@@ -0,0 +1,379 @@
+                case FieldType.List:
+                    return $"{CollectionTypesToQualifiedTypes[Collection.List]}<{GetQualifiedTypeName(field.ListType.InnerType, lookups)}>";
+                case FieldType.Map:
+                    return $"{CollectionTypesToQualifiedTypes[Collection.Map]}<{GetQualifiedTypeName(field.MapType.KeyType, lookups)}, {GetQualifiedTypeName(field.MapType.ValueType, lookups)}>";
+                case FieldType.Singular:
+                    return GetQualifiedTypeName(field.SingularType.Type, lookups);
+                default:
+                    throw new ArgumentOutOfRangeException();
+            }
+        }
+
+        public static string GetNameFromQualifiedName(string qualifiedName)
+        {
+            return qualifiedName.Substring(qualifiedName.LastIndexOf('.') + 1);
+        }
+
+        public static string GetQualifiedTypeName(TypeDescription type)
+        {
+            return "::" + ReplacesDotsWithDoubleColons(type.QualifiedName);
+        }
+
+        public static string GetQualifiedTypeName(string qualifiedName)
+        {
+            return "::" + ReplacesDotsWithDoubleColons(qualifiedName);
+        }
+
+        public static string GetQualifiedTypeName(ValueTypeReference typeRef, Bundle bundle)
+        {
+            switch (typeRef.ValueTypeSelector)
+            {
+                case ValueType.Enum:
+                    return $"::{GetQualifiedTypeFromQualifiedName(typeRef.Enum.QualifiedName)}";
+                case ValueType.Primitive:
+                    return SchemaToCppTypes[typeRef.Primitive];
+                case ValueType.Type:
+                    var qualifiedName = typeRef.Type.QualifiedName;
+                    return bundle.IsNestedType(qualifiedName) ? GetTypeClassDefinitionQualifiedName(qualifiedName, bundle)
+                        : GetQualifiedTypeFromQualifiedName(qualifiedName); ;
+                default:
+                    throw new ArgumentOutOfRangeException();
+            }
+        }
+
+        public static string GetQualifiedTypeName(TypeReference type)
+        {
+            return ReplacesDotsWithDoubleColons(type.QualifiedName);
+        }
+
+        public static string TypeToHeaderFilename(string qualifiedName)
+        {
+            return TypeToFilename(qualifiedName, ".h");
+        }
+
+        public static string TypeToSourceFilename(string qualifiedName)
+        {
+            return TypeToFilename(qualifiedName, ".cpp");
+        }
+
+        // For each const get accessor for each member field in a type, if the type is a primitive we return by value, otherwise by const ref
+        // e.g. double get_my_double_memeber();
+        //      const ::improbable::List<...>& get_my_list_memeber();
+        public static string GetConstAccessorTypeModification(FieldDefinition field, Bundle bundle)
```

Comment:
This function will not add const for String types and TArrays. We will need to add some logic either here or wherever it is used to ensure we get the correct const markup